routing/payment_lifecycle.go

```diff
@@ -30,21 +30,154 @@ type paymentLifecycle struct {
 	shardTracker  shards.ShardTracker
 	timeoutChan   <-chan time.Time
 	currentHeight int32
+	state         *paymentState
+}
+
+// collectResultAsync launches a goroutine that will wait for the result of the
+// given HTLC attempt to be available then handle its result. It will fail the
+// payment with the control tower if a terminal error is encountered.
+//
+// Note: a lock is required on the payment state (p.state) to make sure it
+// stays in consistency within this on call.
+func (p *paymentLifecycle) collectResultAsync(sh *shardHandler,
+	attempt *channeldb.HTLCAttemptInfo) {
+
+	var (
+		wg sync.WaitGroup
+
+		// errToSend is the error to be sent to sh.shardErrors.
+		errToSend error
+	)
+
+	// handleResultErr is a function closure must be called using defer. It
+	// finishes collecting result by updating the payment state and send
+	// the error (or nil) to sh.shardErrors.
+	handleResultErr := func() {
+		// Update the payment state. If an error occurred here, we will
+		// overwrite the errToSend (could be nil atm) and send it to
+		// sh.shardErrors because it's more critical to keep the
+		// payment state intact.
+		if _, err := p.updatePaymentState(); err != nil {
+			log.Errorf("Error updating payment state for "+
+				"shard %v for payment %v: %v",
+				attempt.AttemptID, p.identifier, err)
+
+			errToSend = err
+		}
+
+		// Release the lock.
+		p.state.Unlock()
+
+		// Send the error or quit.
+		select {
+		case sh.shardErrors <- errToSend:
+		case <-p.router.quit:
+		case <-sh.quit:
+		}
+
+		wg.Done()
+	}
+
+	wg.Add(1)
+	go func() {
+		defer handleResultErr()
+
+		// Acquire a lock here because the following functions involve
+		// several db transactions that affect different parts of the
+		// payment state. For instance, a payment is failed in two
+		// steps, 1) fail the HTLC attempt and 2) fail the payment
+		// itself. If, however, another goroutine access the payment
+		// state in-between the steps, the payment state then have info
+		// that's inconsistent.
+		p.state.Lock()
+
+		// Block until the result is available.
+		result, err := sh.collectResult(attempt)
```

Comment:
@halseth Suppose the payment state is, 
```
terminate: false
remainingAmt: 0
numInflightShards: 2
```

There are two steps here, 
1. We fail the attempt inside the `collectResult`, this has the effect of marking one of the `payment.HTLCs` as failed, the payment state then becomes, 
    ```
    terminate: false
    remainingAmt: 5000
    numInflightShards: 1
    ```
2. Then calls `handleSendError` at line 116, which will fail the payment, and the state becomes,
    ```
    terminate: true
    remainingAmt: 5000
    numInflightShards: 1
    ```

Between step 1 and 2, other goroutines may act and change the state.

Reply:
Thanks for the explanation! I'm still trying to understand why changing the state between step 1 and 2 would be a problem though. Payment attempt outcome and overall payment outcome should ideally be decoupled enough for that not to be a problem.

What could happen between 1 and 2 that would lead to unexpected behavior?

Reply:
Thank you for reading this!
Tbh I am too trying to understand its implications. The above case could lead to the situation where we would fire one more shard in the next round of the `lifecycle` loop. Suppose this new one is settled, but it failed to update the payment in db, and the previous step 2 now finished, then we have a payment that's both settled and failed? (Don't know if it's possible tho, still needs to read more code on MMP part).

This is a case when I was refactoring the tests, data race came up because we were updating the payment in one goroutine, while reading it from another. Previously, whenever we construct the payment state, by calling `payment.SentAmt()`, `payment.TerminalInfo()`, etc, there is another goroutine modifying them.

I've also updated this PR, realized that there's a mutex used in control tower when we update the payment. That mutex can be used too when we read the payment state.

Reply:
If the payment has both failed and succeed (which can happen if both happens simultaneously), the success takes precedence: https://github.com/lightningnetwork/lnd/blob/9163d1a16f20b3dad1d23baeb5970cd048fed3a9/channeldb/payments.go#L349-L367

I see what you mean that firing a new shard is unnecessary after the attempt is failed, but before we have marked the payment failed, but it should be fine.

Maybe the data race you saw is a missing mutex within the mock used during the test?

Reply:
I think the mutex is not needed here since it is only used when sending a notification.